static/rison.ts

```diff
@@ -0,0 +1,448 @@
+                return o;
+            },
+            "'": () => {
+                const s = this.string;
+                let i = this.index;
+                let start = i;
+                const segments: string[] = [];
+                let c;
+                while ((c = s.charAt(i++)) !== "'") {
+                    //if (i == s.length) return this.error('unmatched "\'"');
+                    if (!c) this.error('unmatched "\'"');
+                    if (c === '!') {
+                        if (start < i - 1) segments.push(s.slice(start, i - 1));
+                        c = s.charAt(i++);
+                        if ("!'".indexOf(c) >= 0) {
+                            segments.push(c);
+                        } else {
+                            this.error('invalid string escape: "!' + c + '"');
+                        }
+                        start = i;
+                    }
+                }
+                if (start < i - 1) segments.push(s.slice(start, i - 1));
+                this.index = i;
+                return segments.length === 1 ? segments[0] : segments.join('');
+            },
+            // Also any digit.  The statement that follows this table
+            // definition fills in the digits.
+            '-': () => {
+                let s = this.string;
+                let i = this.index;
+                const start = i - 1;
+                let state = 'int';
+                let permittedSigns = '-';
+                const transitions = {
+                    'int+.': 'frac',
+                    'int+e': 'exp',
+                    'frac+e': 'exp',
+                };
+                do {
+                    const c = s.charAt(i++);
+                    if (!c) break;
+                    if ('0' <= c && c <= '9') continue;
+                    if (permittedSigns.indexOf(c) >= 0) {
+                        permittedSigns = '';
+                        continue;
+                    }
+                    state = transitions[state + '+' + c.toLowerCase()];
+                    if (state === 'exp') permittedSigns = '-';
+                } while (state);
+                this.index = --i;
+                s = s.slice(start, i);
+                if (s === '-') this.error('invalid number');
+                return Number(s);
+            },
+        };
+        // copy table['-'] to each of table[i] | i <- '0'..'9':
+        for (let i = 0; i <= 9; i++) this.table[String(i)] = this.table['-'];
+    }
+
+    /**
+     * parse a rison string into a javascript structure.
+     */
+    parse(str: string): JSONValue {
+        this.string = str;
+        this.index = 0;
+        const value = this.readValue();
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+        if (this.next()) this.error("unable to parse string as rison: '" + encode(str) + "'");
+        return value;
+    }
+
+    error(message: string): never {
+        throw new Error('rison parser error: ' + message);
```

Comment:
There was a weird pattern in the original code with `return this.error(...);`, but `this.error` was silent - it would console.log if it could and set `this,message` - then return undefined. I removed that pattern because it seems like a bad idea and makes typing hard.